text/0000-native-lifecycle-mixed-mode.md

```diff
@@ -0,0 +1,152 @@
+- `'reset'` (default, i.e. synthetic mode)
+
+Any other value throws an error at runtime.
+
+### Polyfill behavior
+
+Similar to [`@lwc/synthetic-shadow`](https://www.npmjs.com/package/@lwc/synthetic-shadow) and [`@lwc/aria-reflection`](https://www.npmjs.com/package/@lwc/aria-reflection), we will isolate the code necessary for synthetic lifecycle into a standalone package: `@lwc/synthetic-lifecycle`. The polyfill can be loaded using simply:
+
+```js
+import '@lwc/synthetic-lifecycle';
+```
+
+This allows `@lwc/engine-dom` to run in two modes:
+
+1. **Purely native**. This is the case where the polyfill is not loaded. All components run with native lifecycle.
+2. **Mixed mode**. This is the case where the polyfill _is_ loaded. This is the only case where `lifecycleSupportMode` actually does anything. Components that specify `static lifecycleSupportMode = 'native'` will opt-out of the polyfill.
+
+This has a few advantages:
+
+1. `@lwc/engine-dom` is "pure" and polyfill-free by default, which gives off-core consumers the best possible LWC experience.
+2. It aligns neatly with the existing behavior for LWC's other polyfills.
+3. If consumers do not need the polyfill, then they do not need to load the code for the polyfill (which admittedly is small – roughly 385 bytes minified+gzipped).
+4. It does not enshrine `lwcRuntimeFlags` as a user-facing API surface, which is the current ad-hoc solution to toggling native vs synthetic lifecycle.
+
+### Transitivity
+
+Like shadow DOM mixed mode, lifecycle mixed mode will need some concept of transitivity. In particular, it will enforce the following rules:
+
+1. Any descendants of a native-lifecycle component are forced into native lifecycle mode. (Same as shadow DOM mixed mode.)
+2. Any _slotted content_ of a native-lifecycle component is also forced into native lifecycle mode. (Different from shadow DOM mixed mode.)
+
+The reason for rule #2 is that slots in particular proved to be a source of observable changes for native-lifecycle components. It is simply the case that a slotted component can observe the lifecycle mode of the component it is being slotted into, due to connect/disconnect events triggering downstream callbacks in the tree.
+
+The reason for rule #1 is that, exactly like slotted content, children can observe the mode of their parents.
+
+So rather than trying to emulate synthetic-lifecycle-within-native-lifecycle (which is [a fool's errand](https://github.com/salesforce/lwc/issues/4249)), we should force the entire tree into native lifecycle mode so that `connectedCallback` and `disconnectedCallback` at least consistently fire.
+
+For component authors, this means that they will have to be careful when enabling native lifecycle mode. Similar to native shadow mode, they will have to ensure that their children are all native-lifecycle ready. But in addition, they must communicate to their consumers that any _slotted_ content provided by the consumer is also beholden to this rule.

```

Comment:
This is the part that really makes me hesitate, but I don't see a way around it.

Reply:
> For component authors, this means that they will have to be careful when enabling native lifecycle mode. Similar to native shadow mode, they will have to ensure that their children are all native-lifecycle ready. But in addition, they must communicate to their consumers that any _slotted_ content provided by the consumer is also beholden to this rule.

@nolanlawson Would it make sense to allow the customer to toggle the `lifecycleSupportMode` for an internal component?

I'm thinking of something like base components that have high external customer usage.

Something like this:

```html
<base-component lifecycle-support-mode='native'>
    <c-customer-component></c-customer-component>
</base-component>
```
This would allow internal components to move to native lifecycle but give customer an escape hatch in case of any incompatibilities.

The main drawback I can see is that the internal components would have to design for both synthetic and native lifecycles.

Reply:
No, I am thinking of it more like this:

- Internal components will probably have to support both modes simultaneously (due to downstream impact)
- External components will have more flexibility (due to low-to-zero downstream impact)

This is almost exactly the same situation we have with `shadowSupportMode` BTW.